File: js/f2md/src/source-coverage.ts
/** Deterministic terminal-state accounting for every source discovered by tree conversion. */
import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join, relative, sep } from "node:path";

export const SOURCE_COVERAGE_SCHEMA = "bioxfoundry.source-coverage/v1" as const;
export const SOURCE_STATES = [
  "converted",
  "binary-provenance",
  "excluded-by-policy",
  "unsupported",
  "quarantined",
  "failed",
] as const;

export type SourceCoverageState = typeof SOURCE_STATES[number];
export type TwinRevisionCoverageStatus = "not-evaluated" | "included" | "excluded";

export interface SourceCoverageRecord {
  path: string;
  inputKind: string;
  mediaType: string;
  sourceSha256: string;
  resourceUri: string | null;
  markdownPath: string | null;
  intentUris: string[];
  treeRefs: string[];
  converter: string;
  converterVersion: string;
  state: SourceCoverageState;
  reasonCode: string;
  twinRevisionStatus: TwinRevisionCoverageStatus;
}

export interface SourceCoverageDocument {
  schema: typeof SOURCE_COVERAGE_SCHEMA;
  sourceSnapshotSha256: string;
  coverageSha256: string;
  summary: {
    discovered: number;
    terminal: number;
    byState: Record<SourceCoverageState, number>;
  };
  records: SourceCoverageRecord[];
}

function sha256(value: string | Uint8Array): string {
  return createHash("sha256").update(value).digest("hex");
}

function canonical(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  const object = value as Record<string, unknown>;
  return `{${Object.keys(object).sort().map((key) => `${JSON.stringify(key)}:${canonical(object[key])}`).join(",")}}`;
}

function logicalPath(root: string, path: string): string {
  const value = relative(root, path).split(sep).join("/");
  if (value === ".." || value.startsWith("../") || value.startsWith("/")) {
    throw new Error(`SOURCE_COVERAGE_PATH_ESCAPE:${value}`);
  }
  return value;
}

export async function sourceCoverageRecord(input: {
  root: string;
  path: string;
  inputKind: string;
  mediaType: string;
  state: SourceCoverageState;
  reasonCode: string;
  markdownPath?: string;
  converter?: string;
  converterVersion?: string;
}): Promise<SourceCoverageRecord> {
  const path = logicalPath(input.root, input.path);
  const sourceSha256 = sha256(await readFile(input.path));
  const parent = path.split("/").slice(0, -1).join("/") || ".";
  return {
    path,
    inputKind: input.inputKind,
    mediaType: input.mediaType,
    sourceSha256,
    resourceUri: input.state === "excluded-by-policy" || input.state === "quarantined"
      ? null
      : `urn:subactor:resource:sha256:${sourceSha256}`,
    markdownPath: input.markdownPath?.split(sep).join("/") ?? null,
    intentUris: [],
    treeRefs: [parent],
    converter: input.converter ?? "none",
    converterVersion: input.converterVersion ?? "unknown",
    state: input.state,
    reasonCode: input.reasonCode,
    twinRevisionStatus: "not-evaluated",
  };
}

export function buildSourceCoverage(
  sourceSnapshotSha256: string,
  records: SourceCoverageRecord[],
): SourceCoverageDocument {
  const ordered = [...records].sort((left, right) => Buffer.compare(Buffer.from(left.path), Buffer.from(right.path)));
  const byState = Object.fromEntries(
    SOURCE_STATES.map((state) => [state, ordered.filter((record) => record.state === state).length]),
  ) as Record<SourceCoverageState, number>;
  const terminal = Object.values(byState).reduce((sum, count) => sum + count, 0);
  if (terminal !== ordered.length) {
    throw new Error(`SOURCE_COVERAGE_TERMINAL_MISMATCH:${terminal}:${ordered.length}`);
  }
  const material = {
    schema: SOURCE_COVERAGE_SCHEMA,
    sourceSnapshotSha256,
    summary: { discovered: ordered.length, terminal, byState },
    records: ordered,
  };
  return {
    schema: SOURCE_COVERAGE_SCHEMA,
    sourceSnapshotSha256,
    coverageSha256: sha256(canonical(material)),
    summary: material.summary,
    records: ordered,
  };
}

export function renderSourceCoverageDsl(report: SourceCoverageDocument): string {
  const lines = [
    `SOURCE_COVERAGE ${report.coverageSha256}`,
    `SCHEMA ${report.schema}`,
    `SOURCE_SNAPSHOT ${report.sourceSnapshotSha256}`,
    `DISCOVERED ${report.summary.discovered}`,
    `TERMINAL ${report.summary.terminal}`,
    ...SOURCE_STATES.map((state) => `STATE ${state} ${report.summary.byState[state]}`),
  ];
  for (const record of report.records) {
    lines.push(
      `SOURCE ${JSON.stringify(record.path)}`,
      `  KIND ${JSON.stringify(record.inputKind)}`,
      `  MEDIA_TYPE ${JSON.stringify(record.mediaType)}`,
      `  SOURCE_SHA256 ${record.sourceSha256}`,
      `  RESOURCE_URI ${JSON.stringify(record.resourceUri)}`,
      `  MARKDOWN_PATH ${JSON.stringify(record.markdownPath)}`,
      `  INTENT_URIS ${JSON.stringify(record.intentUris)}`,
      `  TREE_REFS ${JSON.stringify(record.treeRefs)}`,
      `  CONVERTER ${JSON.stringify(record.converter)}`,
      `  CONVERTER_VERSION ${JSON.stringify(record.converterVersion)}`,
      `  TERMINAL_STATE ${record.state}`,
      `  REASON ${record.reasonCode}`,
      `  TWIN_REVISION ${record.twinRevisionStatus}`,
      "END_SOURCE",
    );
  }
  lines.push(
    `RESULT ${report.summary.discovered === report.summary.terminal ? "COMPLETE" : "INCOMPLETE"}`,
    "END_SOURCE_COVERAGE",
  );
  return `${lines.join("\n")}\n`;
}

export async function writeSourceCoverage(root: string, report: SourceCoverageDocument): Promise<boolean> {
  await mkdir(root, { recursive: true });
  const jsonPath = join(root, "source-coverage.json");
  const dslPath = join(root, "source-coverage.dsl");
  const jsonBody = `${JSON.stringify(report, null, 2)}\n`;
  const dslBody = renderSourceCoverageDsl(report);
  const same = async (path: string, body: string): Promise<boolean> =>
    readFile(path, "utf8").then((existing) => existing === body, () => false);
  const sameJson = await same(jsonPath, jsonBody);
  const sameDsl = await same(dslPath, dslBody);
  if (!sameJson) await writeFile(jsonPath, jsonBody);
  if (!sameDsl) await writeFile(dslPath, dslBody);
  return sameJson && sameDsl;
}

File: py/f2md/src/f2md/source_coverage.py
"""Deterministic terminal-state accounting for an f2md source tree.

The report is a file contract, not runtime state. It contains no timestamps or host-specific root
paths, so an unchanged tree produces byte-identical JSON and DSL on every run. Whether those bytes
were already present is returned separately to the caller as ``coverage_no_change``.
"""

from __future__ import annotations

import hashlib
import json
import os
from pathlib import Path
from typing import Any, Dict, Sequence

SOURCE_COVERAGE_SCHEMA = "bioxfoundry.source-coverage/v1"
SOURCE_STATES = (
    "converted",
    "binary-provenance",
    "excluded-by-policy",
    "unsupported",
    "quarantined",
    "failed",
)


def file_sha256(path: str) -> str:
    digest = hashlib.sha256()
    with open(path, "rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def logical_path(root: str, path: str) -> str:
    relative = os.path.relpath(path, root).replace(os.sep, "/")
    if relative == ".." or relative.startswith("../") or relative.startswith("/"):
        raise ValueError(f"SOURCE_COVERAGE_PATH_ESCAPE:{relative}")
    return relative


def source_record(
    *,
    root: str,
    path: str,
    input_kind: str,
    media_type: str,
    state: str,
    reason_code: str,
    markdown_path: str | None = None,
    converter: str = "none",
    converter_version: str = "unknown",
) -> Dict[str, Any]:
    if state not in SOURCE_STATES:
        raise ValueError(f"SOURCE_COVERAGE_STATE_INVALID:{state}")
    source_hash = file_sha256(path)
    relative = logical_path(root, path)
    parent = str(Path(relative).parent).replace(os.sep, "/")
    return {
        "path": relative,
        "inputKind": input_kind,
        "mediaType": media_type,
        "sourceSha256": source_hash,
        "resourceUri": f"urn:subactor:resource:sha256:{source_hash}"
        if state not in ("excluded-by-policy", "quarantined")
        else None,
        "markdownPath": markdown_path.replace(os.sep, "/") if markdown_path else None,
        "intentUris": [],
        "treeRefs": [parent if parent not in ("", ".") else "."],
        "converter": converter,
        "converterVersion": converter_version,
        "state": state,
        "reasonCode": reason_code,
        "twinRevisionStatus": "not-evaluated",
    }


def build_source_coverage(source_snapshot_sha256: str, records: Sequence[Dict[str, Any]]) -> Dict[str, Any]:
    ordered = sorted(
        (dict(record) for record in records),
        key=lambda record: str(record["path"]).encode("utf-8"),
    )
    by_state = {state: sum(record.get("state") == state for record in ordered) for state in SOURCE_STATES}
    terminal = sum(by_state.values())
    if terminal != len(ordered):
        raise ValueError(f"SOURCE_COVERAGE_TERMINAL_MISMATCH:{terminal}:{len(ordered)}")
    material = {
        "schema": SOURCE_COVERAGE_SCHEMA,
        "sourceSnapshotSha256": source_snapshot_sha256,
        "summary": {"discovered": len(ordered), "terminal": terminal, "byState": by_state},
        "records": ordered,
    }
    coverage_hash = hashlib.sha256(
        json.dumps(material, ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode("utf-8")
    ).hexdigest()
    return {
        "schema": SOURCE_COVERAGE_SCHEMA,
        "sourceSnapshotSha256": source_snapshot_sha256,
        "coverageSha256": coverage_hash,
        "summary": material["summary"],
        "records": ordered,
    }


def render_source_coverage_dsl(report: Dict[str, Any]) -> str:
    summary = report["summary"]
    lines = [
        f"SOURCE_COVERAGE {report['coverageSha256']}",
        f"SCHEMA {report['schema']}",
        f"SOURCE_SNAPSHOT {report['sourceSnapshotSha256']}",
        f"DISCOVERED {summary['discovered']}",
        f"TERMINAL {summary['terminal']}",
    ]
    for state in SOURCE_STATES:
        lines.append(f"STATE {state} {summary['byState'][state]}")
    for record in report["records"]:
        lines.extend(
            [
                f"SOURCE {json.dumps(record['path'], ensure_ascii=False)}",
                f"  KIND {json.dumps(record['inputKind'], ensure_ascii=False)}",
                f"  MEDIA_TYPE {json.dumps(record['mediaType'], ensure_ascii=False)}",
                f"  SOURCE_SHA256 {record['sourceSha256']}",
                f"  RESOURCE_URI {json.dumps(record['resourceUri'], ensure_ascii=False)}",
                f"  MARKDOWN_PATH {json.dumps(record['markdownPath'], ensure_ascii=False)}",
                f"  INTENT_URIS {json.dumps(record['intentUris'], ensure_ascii=False)}",
                f"  TREE_REFS {json.dumps(record['treeRefs'], ensure_ascii=False)}",
                f"  CONVERTER {json.dumps(record['converter'], ensure_ascii=False)}",
                f"  CONVERTER_VERSION {json.dumps(record['converterVersion'], ensure_ascii=False)}",
                f"  TERMINAL_STATE {record['state']}",
                f"  REASON {record['reasonCode']}",
                f"  TWIN_REVISION {record['twinRevisionStatus']}",
                "END_SOURCE",
            ]
        )
    complete = summary["discovered"] == summary["terminal"]
    lines.extend([f"RESULT {'COMPLETE' if complete else 'INCOMPLETE'}", "END_SOURCE_COVERAGE"])
    return "\n".join(lines) + "\n"


def write_source_coverage(root: str, report: Dict[str, Any]) -> bool:
    """Write JSON and DSL when changed; return whether both exact bytes already existed."""
    os.makedirs(root, exist_ok=True)
    json_path = os.path.join(root, "source-coverage.json")
    dsl_path = os.path.join(root, "source-coverage.dsl")
    json_body = json.dumps(report, ensure_ascii=False, indent=2) + "\n"
    dsl_body = render_source_coverage_dsl(report)

    def unchanged(path: str, body: str) -> bool:
        try:
            return Path(path).read_text(encoding="utf-8") == body
        except (FileNotFoundError, UnicodeDecodeError):
            return False

    no_change = unchanged(json_path, json_body) and unchanged(dsl_path, dsl_body)
    if not unchanged(json_path, json_body):
        Path(json_path).write_text(json_body, encoding="utf-8")
    if not unchanged(dsl_path, dsl_body):
        Path(dsl_path).write_text(dsl_body, encoding="utf-8")
    return no_change
